services/scabbard/libscabbard/src/store/lmdb/mod.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright 2018-2021 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! LMDB-backed CommitHashStore implementations.
+
+use splinter::error::{InternalError, InvalidArgumentError, InvalidStateError};
+use transact::database::{lmdb::LmdbDatabase, Database, DatabaseError};
+
+use crate::hex;
+
+use super::{CommitHashStore, CommitHashStoreError, ServiceId};
+
+const CURRENT_STATE_ROOT_INDEX: &str = "current_state_root";
+
+/// Provides commit log storage using an LMDB database in a legacy configuration.
+///
+/// The legacy database configuration requires a index "current_state_root" configured on the
+/// database instance.  This is expected externally. It also doesn't support multiple services per
+/// store, as it expects a unique DB instance per store.
+pub struct LmdbCommitHashStore {
+    db: LmdbDatabase,
+}
+
+impl LmdbCommitHashStore {
+    /// Constructs a new commit log store around an LMDB database instance.
+    pub fn new(db: LmdbDatabase) -> Self {
+        Self { db }
+    }
+}
+
+impl CommitHashStore for LmdbCommitHashStore {
```

Comment:
Because this implementation ignores ServiceId, it does not fully implement the intent of the trait. This would require higher-level code to include conditionals based on implementation.

If we keep this approach, maybe we should add a service_id argument to new() and then make sure the service id passed on these calls matches it.

Another approach would be to let this implementation handle more than a single lmbd database instance. If this store were filesystem-aware and knew where the lmdb database files resided, this implementation could instantiate LmdbDatabase when it needed it. This could also be split out into a factory pattern if we wanted something else to answer "return the LmdbDatabase for this service id".

Reply:
I have simplified the trait remove the use the ServiceId argument.  Instances will be localized to a specific circuit/service ID pair